File: 2017/advent1_b.hs
module Main where

summarise _ [] s = s
summarise d [x] s = summarise d [] (if x == d then s+x else s)
summarise d (x:y:xs) s = summarise d (y:xs) (if x == y then s+x else s)


summDist [] _ s = s
summDist (x:xs) (y:ys) s = summDist xs ys (if x ==y then s+x else s)


main = do
    input <- getLine

    let digits = map (\d -> read [d] :: Int) input
    let shift = length digits `div` 2
    let (h, t) = splitAt shift digits

    print $ summDist digits (t++h) 0

File: 2017/advent1_b.rs
#![feature(slice_rotate)]

use std::io;


fn summarise(digits: Vec<u32>, paired: Vec<u32>) -> u32 {

    digits.iter().zip(paired).fold(0, |sum, (&x, y)| if x == y { sum + x } else { sum })
}

fn main() {

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();  // should be no biggie :)

    let digits : Vec<u32> = input.chars().map(|c| c.to_digit(10).unwrap()).collect();
    let split = digits.len() / 2;

    let mut paired : Vec<u32> = digits.clone();
    paired.rotate(split);

    let s = summarise(digits, paired);

    println!("sum is {}", s);
}
